fix gotestsum.installer installing wrong version

thaJeztah opened the pull request:
When using go modules, `go build` will always fetch the latest
version of the package, so ignores the version we previously `go get`'d.

Instead of running `go get` and `go build` separately, this patch uses
`go get` (without the `-d` option) to do it all in one step.

Given that this binary is only used for testing, and only used inside the
Dockerfile, we should consider inlining this step in the Dockerfile itself,
but keeping that separate for now.

thaJeztah committed:
fix gotestsum.installer installing wrong version

When using go modules, `go build` will always fetch the latest
version of the package, so ignores the version we previously `go get`'d.

Instead of running `go get` and `go build` separately, this patch uses
`go get` (without the `-d` option) to do it all in one step.

Given that this binary is only used for testing, and only used inside the
Dockerfile, we should consider inlining this step in the Dockerfile itself,
but keeping that separate for now.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
@SamWhited @AkihiroSuda @cpuguy83 PTAL

SamWhited reviewed (approved):
LGTM

thaJeztah commented:
Argh. Windows script also has the same problem

```
cannot find module providing package gotest.tools/gotestsum: working directory is not part of a module
gotestsum build failed...
At line:1 char:533
+ ...  if ($LASTEXITCODE -ne 0) { Throw 'gotestsum build failed...'; } Writ ...
+                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (gotestsum build failed...:Str 
   ing) [], RuntimeException
    + FullyQualifiedErrorId : gotestsum build failed...
 



ERROR: Failed 'ERROR: Failed to build image from Dockerfile.windows' at 06/24/2020 15:28:34
At D:\gopath\src\github.com\docker\docker\hack\ci\windows.ps1:483 char:12
+            Throw "ERROR: Failed to build image from Dockerfile.window ...
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Let me push another commit to fix that one (keeping it in a separate commit as 19.03 does not yet have that part, so easier to backport then)

thaJeztah committed:
Dockerfile.windows: fix gotestsum.installer installing wrong version

When using go modules, `go build` will always fetch the latest
version of the package, so ignores the version we previously `go get`'d.

Instead of running `go get` and `go build` separately, this patch uses
`go get` (without the `-d` option) to do it all in one step.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
CI is failing on a flaky test; https://github.com/moby/moby/issues/41138

thaJeztah commented:
All green now 👍

tianon commented on Dockerfile.windows:
```diff
@@ -262,15 +262,12 @@ RUN `
+    $Env:GO111MODULE = 'on'; `
+    $tmpGobin = "${Env:GOBIN_TMP}"; `
+    $Env:GOBIN = """${Env:GOPATH}`\bin"""; `
+    &go get -buildmode=exe "gotest.tools/gotestsum@${Env:GOTESTSUM_COMMIT}"; `
```
Doesn't this one need `-ldflags` to set `main.version` too?

thaJeztah replied:
Ah, yes, perhaps we should add it there (we didn't do that before, it was mostly when I tried checking if it's the right version)

thaJeztah replied:
or perhaps I should remove it from the other one, as we don't really use it 🤔

let me know if you feel strongly about one or the other

tianon replied:
I don't really feel strongly either way -- just that we should try to keep them consistent. :sweat_smile:

I'm personally fine with this binary not knowing what version it is, but I don't run it very often, so I think I'm the wrong one to opine. :shrug: :heart:

cpuguy83 replied:
Agree, let's just be consistent.

thaJeztah committed:
hack/install: build gotestsum without -buildmode=pie

No need for this binary as it's only used in tests.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

tianon reviewed (approved):
LGTM

thaJeztah commented:
Thanks @tianon 

@cpuguy83 PTAL: I removed the ldflags, and also added a commit to remove `$GO_BUILDMODE`, as I don't think we need `--buildmode=pie` for this binary (likely not for others as well, but didn't look yet)

cpuguy83 reviewed (approved):
LGTM
